### PR TITLE
xproj / csproj interoperability

### PR DESCRIPTION
**Pull request by cdmihai**
This refers to the scenario where VS drives the build and there are xproj projects referencing csproj projects.

When a project.json project references a csproj project, nuget cannot follow the dependency path toward the csproj. Those dependencies are only available at build time. Therefore, there will be a special msbuild target that generates a fragment lock file describing all the csproj projects that an xproj depends upon.

By convention, the fragment lock file is placed besides the master lock file. If the fragment is there, LockfileReader merges with it.

Specs: https://github.com/dotnet/cli/blob/troy/tooling/Documentation/specs/xproj2csproj.md

There's no automated integration tests yet, until we decide where / how to put them: #2123 

Overview of the changes:
- CLI can merge lock files
- Introduced new MSBuildProjectDescription
- Project Model Server knows about MSBuildProjectDescription

---
This change is [Reviewable](https://reviewable.io/reviews/dotnet/cli/2124)

### CODE REVIEW

**Comment by cdmihai** [comment]
@abpiskunov @davidfowl @troydai @anurse

File: test/dotnet-projectmodel-server.Tests/DthTests.cs [hunk 1]
```diff
@@ -412,5 +413,78 @@ public void WaitForLockFileReleased(int occupyFileFor, bool expectSuccess)
...
+                    
+                    Assert.True(classLibraries.Contains(name));
+                    projectRef.AssertProperty<string>("Path", path => path.Contains(Path.Combine("BasicCase01", name)))
+                              .AssertProperty("MSBuildProjectPath", AssertPathsEqual(Path.Combine("..", "..", name, $"{name}.csproj")));                    
+                }
+                
+                var fileReferences = references.RetrievePropertyAs<JArray>("FileReferences")
+                                               .Select(each => each.Value<string>())
+                                               .ToArray();
+                Assert.Equal(7, fileReferences.Length);
```
Comment (cdmihai) [inline test/dotnet-projectmodel-server.Tests/DthTests.cs]: @troydai 

Is this correct? That test was failing with expected:3; actual:7

File: src/Microsoft.DotNet.ProjectModel/TargetLibraryWithAssets.cs [hunk 2]
```diff
@@ -0,0 +1,47 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System.Collections.Generic;
+using Microsoft.DotNet.ProjectModel.Graph;
+using NuGet.Frameworks;
+
+namespace Microsoft.DotNet.ProjectModel
+{
+    public abstract class TargetLibraryWithAssets : LibraryDescription
```
Comment (davidfowl) [inline src/Microsoft.DotNet.ProjectModel/TargetLibraryWithAssets.cs]: This name is pretty bad, what about LockFileTargetDescription

File: src/dotnet/commands/dotnet-projectmodel-server/InternalModels/ProjectContextSnapshot.cs [hunk 3]
```diff
@@ -51,15 +51,13 @@ public static ProjectContextSnapshot Create(ProjectContext context, string confi
                 var description = DependencyDescription.Create(export.Library, diagnostics, allExports);
                 allDependencies[description.Name] = description;
 
-                var projectDescription = export.Library as ProjectDescription;
-                if (projectDescription != null)
+                var projectReferene = ProjectReferenceDescription.Create(export.Library);
```
Comment (davidfowl) [inline src/dotnet/commands/dotnet-projectmodel-server/InternalModels/ProjectContextSnapshot.cs]: stange to see Create return null.

File: src/dotnet/commands/dotnet-publish/PublishCommand.cs [hunk 4]
```diff
@@ -207,6 +181,40 @@ private bool PublishProjectContext(ProjectContext context, string buildBasePath,
             return true;
         }
 
+        private Boolean InvokeBuildOnProject(ProjectContext context, String buildBasePath, String configuration)
```
Comment (davidfowl) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: Boolean? change to `bool`, also change uppercase `String` to lowercase `string`

Comment (cdmihai) [inline src/dotnet/commands/dotnet-publish/PublishCommand.cs]: Good old extract method code generation :) 

---
In reply to: [57806699](https://github.com/dotnet/cli/pull/2124#discussion_r57806699) [](ancestors = 57806699)

File: test/dotnet-projectmodel-server.Tests/DthTests.cs (hunk 1 above)
Comment (troydai) [inline test/dotnet-projectmodel-server.Tests/DthTests.cs]: Should be 3 ... why was it 7? My bad.

File: TestAssets/LockFileSamples/ExportFileSamples/valid/project.lock.json [hunk 5]
```diff
@@ -0,0 +1,61 @@
+{
```
Comment (piotrpMSFT) [inline TestAssets/LockFileSamples/ExportFileSamples/valid/project.lock.json]: I'm not sure 'Samples' is the right word in the directory names. How about just LockFiles? Also, ExportFile~~Sample~~s seem like peers to LockFiles, not a specialization of them.

File: TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/ClassLibrary1/Class1.cs [hunk 6]
```diff
@@ -0,0 +1,12 @@
+﻿using System;
```
Comment (piotrpMSFT) [inline TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/ClassLibrary1/Class1.cs]: This should be more expressive than BasicCase01. What does this project represent?

File: TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/src/ConsoleApp13/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Reflection;
```
Comment (piotrpMSFT) [inline TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/src/ConsoleApp13/Properties/AssemblyInfo.cs]: Why ConsoleApp**13**?

File: src/Microsoft.DotNet.ProjectModel/Graph/LibraryType.cs [hunk 8]
```diff
@@ -11,11 +11,13 @@ public struct LibraryType : IEquatable<LibraryType>
         public static readonly LibraryType Package = new LibraryType(nameof(Package));
         public static readonly LibraryType Project = new LibraryType(nameof(Project));
         public static readonly LibraryType ReferenceAssembly = new LibraryType(nameof(ReferenceAssembly));
+        public static readonly LibraryType MSBuildProject = new LibraryType(nameof(MSBuildProject));
 
         // Default value
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.ProjectModel/Graph/LibraryType.cs]: Kill this comment

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+
+namespace Microsoft.DotNet.ProjectModel.Tests
+{
+    public class LockFilePatchingTests : TestBase
+    {
+
+        private string ExportFilesRoot=> Path.Combine(RepoRoot, "TestAssets", "LockFileSamples", "ExportFileSamples");
+
+        [Fact]
+        public void TestValidPatching()
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: The relationship between this test and Patching is very unclear. This test seems to be `WhenLockFileIsValidThenExportFileRepresentsItsContents`. Is patching just an implementation detail that is not being directly evaluated?

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+            var lockFile = LockFileReader.Read(lockFilePath);
+
+            var exportFile = lockFile.ExportFile;
+
+            exportFile.Should().NotBeNull();
+            exportFile.Exports.Count.Should().Be(3);
+            exportFile.Exports.Should().OnlyHaveUniqueItems();
+
+            // check export structure
+            for (int i = 0; i < 3; i++)
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: This is imposing a requirement that order be maintained. I get that this makes the test a bit easier to write, but we shouldn't suggest that this is requirement... unless it is?

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 11]
```diff
@@ -0,0 +1,101 @@
...
+                var export = exportFile.Exports.ToList().ElementAt(i);
+
+                export.TargetFramework.Should().NotBeNull();
+
+                AssertTargetLibrary(i + 1, export);
+            }
+
+            lockFile.Targets.Count.Should().Be(3);
+
+            // check lock file structure is similar to export structure
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: This seems like an independent test. We're not validating any relationship across these collections, we are independently validating each collection's contents. Can this be split up?

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 12]
```diff
@@ -0,0 +1,101 @@
...
+                }
+            }
+        }
+
+        [Fact]
+        public void TestMissingExportFileThrows()
+        {
+            var lockFilePath = GetLockFilePath("invalid_nofragment");
+
+            Assert.Throws<FileFormatException>(() => LockFileReader.Read(lockFilePath));
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: Can we validate that the exception message is useful? This is not localizable text, so it's safe to test.

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        [Fact]
+        public void TestMissingExportsThrow()
+        {
+            var lockFilePath = GetLockFilePath("invalid_missing-exports");
+
+            Assert.Throws<FileFormatException>(() => LockFileReader.Read(lockFilePath));
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: and here

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 14]
```diff
@@ -0,0 +1,101 @@
...
+        }
+
+        [Fact]
+        public void TestMissmatchingFileVersionsThrows()
+        {
+            var lockFilePath = GetLockFilePath("invalid_missmatching-versions");
+
+            Assert.Throws<FileFormatException>(() => LockFileReader.Read(lockFilePath));
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: And here

File: test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs [hunk 15]
```diff
@@ -0,0 +1,101 @@
...
+
+            export.RuntimeAssemblies.Count.Should().Be(1);
+            export.RuntimeAssemblies.ElementAt(0).Path.Should().Be(dll);
+        }
+
+        private string GetLockFilePath(string exportSample)
+        {
+            return Path.Combine(ExportFilesRoot, exportSample, "project.lock.json");
+        }
+    }
```
Comment (piotrpMSFT) [inline test/Microsoft.DotNet.ProjectModel.Tests/LockFilePatchingTests.cs]: The name of this file doesn't convey to me what intent is being tested. I think we're actually testing the LockFileReader and its ability to parse a project.lock.json. That should be made more clear.

File: test/dotnet-projectmodel-server.Tests/DthTests.cs [hunk 16]
```diff
@@ -412,5 +413,78 @@ public void WaitForLockFileReleased(int occupyFileFor, bool expectSuccess)
...
+
+            using (var server = new DthTestServer(_loggerFactory))
+            using (var client = new DthTestClient(server, _loggerFactory))
+            {
+                client.Initialize(testProject);
+                var messages = client.DrainAllMessages();
+                
+                var classLibraries = new HashSet<string>(new string[] { "ClassLibrary1", "ClassLibrary2", "ClassLibrary3" });
+                var dependencies = messages.RetrieveSingleMessage(MessageTypes.Dependencies);
+                foreach (var each in classLibraries)
```
Comment (piotrpMSFT) [inline test/dotnet-projectmodel-server.Tests/DthTests.cs]: var classLibrary in classLibraries

File: src/Microsoft.DotNet.ProjectModel/Compilation/LibraryExporter.cs [hunk 17]
```diff
@@ -154,22 +155,26 @@ private LibraryExport GetExport(LibraryDescription library)
...
-            builder.WithSourceReferences(GetSharedSources(package));
-            builder.WithAnalyzerReference(GetAnalyzerReferences(package));
-
-            if (package.ContentFiles.Any())
+            var builder = LibraryExportBuilder.Create(library);
+            builder.AddNativeLibraryGroup(new LibraryAssetGroup(PopulateAssets(library, library.NativeLibraries)));
+            builder.AddRuntimeAssemblyGroup(new LibraryAssetGroup(PopulateAssets(library, library.RuntimeAssemblies)));
+            builder.WithCompilationAssemblies(PopulateAssets(library, library.CompileTimeAssemblies));
+
+            if (library.Identity.Type.Equals(LibraryType.Package))
```
Comment (piotrpMSFT) [inline src/Microsoft.DotNet.ProjectModel/Compilation/LibraryExporter.cs]: This condition is untested

File: TestAssets/LockFileSamples/ExportFileSamples/valid/project.fragment.lock.json [hunk 18]
```diff
@@ -0,0 +1,35 @@
+{
```
Comment (piotrpMSFT) [inline TestAssets/LockFileSamples/ExportFileSamples/valid/project.fragment.lock.json]: The use of `valid` and `invalid` seems inappropriate for test assets. These are projects that exhibit some characteristics. Those characteristics' validity is really defined in tests. I'd have an easier time debugging tests if the projects describe the scenarios they are representing and the test code told me if/why valid.

**Comment by cdmihai** [comment]
@dotnet-bot test this please

File: TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/src/ConsoleApp13/Properties/AssemblyInfo.cs (hunk 7 above)
Comment (cdmihai) [inline TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/src/ConsoleApp13/Properties/AssemblyInfo.cs]: We copied over a sample @abpiskunov gave us, and then trimmed it down. :)

Comment (troydai) [inline TestAssets/ProjectModelServer/MSBuildReferencesProjects/BasicCase01/src/ConsoleApp13/Properties/AssemblyInfo.cs]: That's the sample, I got. We can change the name, no problem.

File: src/Microsoft.DotNet.ProjectModel/TargetLibraryWithAssets.cs (hunk 2 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.ProjectModel/TargetLibraryWithAssets.cs]: I like the current name. It means that this class represents target libraries for which the asset fields exist and make sense. This is the case for msbuild and packages. Library target objects representing project.json projects do not have those asset fields populated.

What about LibraryDescriptionWithAssets?

File: TestAssets/LockFileSamples/ExportFileSamples/valid/project.lock.json (hunk 5 above)
Comment (cdmihai) [inline TestAssets/LockFileSamples/ExportFileSamples/valid/project.lock.json]: Export files appear near a specialization of Lock files. They don't exist without lock files, and when they do exist, the lock files are special, in the sense that they have holes in them

I'll remove the samples bit.

File: src/Microsoft.DotNet.ProjectModel/Compilation/LibraryExporter.cs (hunk 17 above)
Comment (cdmihai) [inline src/Microsoft.DotNet.ProjectModel/Compilation/LibraryExporter.cs]: All tests that request LibraryExports from nuget packages exercise the branch getting taken.

The branch getting not taken can't be exercised until we decide on #2123

**Pull request merged**